File: Data Structures & Algorithms/products-of-array-discluding-self/submission-0.java
class Solution {
    public int[] productExceptSelf(int[] nums) {
        int[] exclusivePrefixArr = new int[nums.length];
        exclusivePrefixArr[0] = 1;
        for(int i = 1; i<nums.length; i++){
            exclusivePrefixArr[i] = exclusivePrefixArr[i-1] * nums[i-1];
        }

        int runningProduct = 1;
        for(int i = nums.length - 1; i>=0; i--){
            exclusivePrefixArr[i] = exclusivePrefixArr[i] * runningProduct;
            runningProduct *= nums[i];
        }

        return exclusivePrefixArr;
    }
}  

File: Data Structures & Algorithms/combination-target-sum/submission-0.js
class Solution {
    /**
     * state -> where I currently stand & conbination
     * 
     */
    dfs(nums, target, i, combination, currSum ,result){
        // if currSum is equeal to target -> I push the combination to the result
        // if currSum exceeds, I return without anything

        if (i >= nums.length) {
            return;
        }
        
        if(currSum > target){
            return;
        }

        if(currSum === target){
            result.push([...combination]);
            return;
        }

        // choose

        combination.push(nums[i]);

        // recurse

        this.dfs(nums, target, i, combination, currSum + nums[i], result);

        // remove choice

        combination.pop();

        // skip and recurse

        this.dfs(nums, target, i+1, combination, currSum, result);

    }


    /**
     * @param {number[]} nums
     * @param {number} target
     * @returns {number[][]}
     */
    combinationSum(nums, target) {
        const result = [];
        this.dfs(nums, target, 0, [], 0, result);
        return result;
    }
}
